casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -461,25 +460,35 @@ class Running[F[_]
       handleForkChoiceTipRequest(peer)(casper)
     case abr: ApprovedBlockRequest =>
       for {
-        approvedBlock <- if (abr.trimState) for {
-                          lfBlock <- LastFinalizedStorage[F]
-                                      .get(approvedBlock.candidate.block)
-                                      .flatMap(BlockStore[F].getUnsafe)
-                          // Each approved block should be justified by validators signatures
-                          // ATM we have signatures only for genesis approved block - we also have to have a procedure
-                          // for gathering signatures for each approved block post genesis.
-                          // Now new node have to trust bootstrap if it wants to trim state when connecting to the network.
-                          // TODO We need signatures of Validators supporting this block
-                          lastApprovedBlock = ApprovedBlock(
-                            ApprovedBlockCandidate(lfBlock, 0),
-                            List.empty
-                          )
-                        } yield lastApprovedBlock
+        lfBlockHashOpt <- LastFinalizedStorage[F].get()
+
+        // Create approved block from last finalized block
+        lastFinalizedBlock = for {
+          lfBlockHash <- lfBlockHashOpt.liftTo[F](
+                          new Exception(s"Last finalized block hash not available.")
+                        )
+          lfBlock <- BlockStore[F].getUnsafe(lfBlockHash)
+
+          // Each approved block should be justified by validators signatures
+          // ATM we have signatures only for genesis approved block - we also have to have a procedure
+          // for gathering signatures for each approved block post genesis.
+          // Now new node have to trust bootstrap if it wants to trim state when connecting to the network.
+          // TODO We need signatures of Validators supporting this block
```

Comment:
Question. With the feature above, Is last finalized block the same as the approved block?
Last finalized should be justified by the blocks built above it which got the signature from most validator on each block.

I am also thinking that we should gather signatures but just for epoch change block.

Reply:
@zsluedem yes, I think you are right, that why I was pushing name not `last finalized block`, but `approved block`.

`Last finalized should be justified by the blocks built above it which got the signature from most validator on each block.` yes, this makes sense. If approved block (last finalized block) received will be rejected by validators (noone will build on top of that) -  it will be revealed pretty soon that this is a invalid/malicious approved block. I think it's safe to trust bootstrap in any case, even on epoch change.